File: services/k9-runtime/k9_models.py
"""Resolves K9 model locations from config/k9-models.json.

The Python counterpart of scripts/k9-models.mjs and the api-server's
lib/k9-models.ts. Runtime code must get every model, weight and code path from
here — never hard-code a model path.
"""

from __future__ import annotations

import fnmatch
import json
import os
from pathlib import Path
from typing import Any, Mapping

COMPLETE_MARKER = ".k9_complete"
_HERE = Path(__file__).resolve()


def default_config_path() -> Path:
    explicit = os.environ.get("K9_MODELS_CONFIG")
    if explicit:
        return Path(explicit).resolve()
    candidates = (
        _HERE.parents[2] / "config" / "k9-models.json",  # repo: services/k9-runtime/k9_models.py
        _HERE.parents[1] / "config" / "k9-models.json",  # desktop app: resources/runtime/k9_models.py
    )
    return next((c for c in candidates if c.exists()), candidates[0])


def _has_incomplete_downloads(folder: Path) -> bool:
    cache = folder / ".cache" / "huggingface" / "download"
    return cache.exists() and any(p.name.endswith(".incomplete") for p in cache.rglob("*"))


class Registry:
    def __init__(self, path: str | Path | None = None, env: Mapping[str, str] | None = None):
        self.path = Path(path).resolve() if path else default_config_path()
        self.env = os.environ if env is None else env
        self.config: dict[str, Any] = json.loads(self.path.read_text(encoding="utf-8"))
        self.roots = {
            name: Path((root.get("env") and self.env.get(root["env"])) or root["path"])
            for name, root in self.config["roots"].items()
        }

    @property
    def models(self) -> dict[str, dict[str, Any]]:
        return self.config["models"]

    def runtime(self, key: str, default: Any = None) -> Any:
        return self.config.get("runtime", {}).get(key, default)

    def entry(self, model_id: str) -> dict[str, Any]:
        try:
            return self.models[model_id]
        except KeyError:
            raise KeyError(f"{model_id} is not in the K9 model registry ({self.path})") from None

    def path_of(self, model_id: str) -> Path:
        entry = self.entry(model_id)
        return self.roots[entry.get("root", "k9")].joinpath(*entry["path"].split("/"))

    def weights_of(self, model_id: str) -> Path:
        """The file a runtime loads: the entry's `weights` inside its folder, or the entry itself."""
        entry = self.entry(model_id)
        base = self.path_of(model_id)
        return base.joinpath(*entry["weights"].split("/")) if entry.get("weights") else base

    def code_of(self, model_id: str) -> Path:
        entry = self.entry(model_id)
        if not entry.get("code"):
            raise KeyError(f"{model_id} has no `code` path in the K9 model registry")
        return self.path_of(model_id).joinpath(*entry["code"].split("/"))

    def status(self, model_id: str) -> str:
        """"ready" | "partial" | "missing" — the same rules as scripts/k9-models.mjs."""
        entry = self.entry(model_id)
        path = self.path_of(model_id)
        if not path.exists():
            return "missing"
        kind = entry["kind"]
        if kind == "file":
            return "ready" if path.stat().st_size > 0 else "partial"
        if kind == "repo":
            ready = (path / ".git").exists()
        elif kind == "files":
            ready = all(path.joinpath(*f.split("/")).exists() for f in entry["files"])
        else:
            names = os.listdir(path)
            expected = all(
                any(fnmatch.fnmatch(name.lower(), pattern.lower()) for name in names)
                for pattern in entry.get("expect", [])
            )
            if entry["source"]["type"] == "external":
                ready = bool(names) and expected and not _has_incomplete_downloads(path)
            else:
                ready = COMPLETE_MARKER in names and expected
        if ready and entry.get("weights") and not self.weights_of(model_id).exists():
            return "partial"
        if ready and entry.get("code") and not self.code_of(model_id).exists():
            return "partial"
        return "ready" if ready else "partial"

File: artifacts/api-server/src/lib/k9-models.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";

// Resolves K9 model locations from the registry at config/k9-models.json —
// the api-server counterpart of scripts/k9-models.mjs. Never hard-code model
// paths in routes or workers; read them through here.

export type K9ModelKind = "file" | "dir" | "repo" | "files";
export type K9ModelStatus = "ready" | "partial" | "missing";
export type K9RootName = "base" | "k9";

export type K9ModelEntry = {
  category: string;
  role: string;
  root?: K9RootName;
  path: string;
  kind: K9ModelKind;
  required: boolean;
  files?: string[];
  expect?: string[];
  gated?: boolean;
  license?: string;
  note?: string;
  source: { type: string; repo?: string; url?: string; id?: string; from?: string; note?: string };
  ollama?: { name: string; quantize?: string; parameters?: Record<string, unknown> };
};

export type K9ModelsConfig = {
  version: number;
  roots: Record<K9RootName, { path: string; env?: string }>;
  runtime?: {
    ollama?: {
      text_model?: string;
      text_fallbacks?: string[];
      vision_model?: string;
      vision_fallbacks?: string[];
    };
    k9_runtime?: { host?: string; port?: number };
    python?: { executable?: string; env?: string };
  };
  models: Record<string, K9ModelEntry>;
};

function ollamaNames(config: K9ModelsConfig, ids: (string | undefined)[]): string[] {
  const names = ids
    .filter((id): id is string => !!id)
    .map((id) => config.models[id]?.ollama?.name)
    .filter((name): name is string => !!name);
  return [...new Set(names)];
}

/** Ollama model names for K9's text brain: the registry default first, then its fallbacks. */
export function k9TextModelNames(config: K9ModelsConfig): string[] {
  const runtime = config.runtime?.ollama;
  return ollamaNames(config, [runtime?.text_model, ...(runtime?.text_fallbacks ?? [])]);
}

/**
 * Ollama model names that can read an image — the paper reader and any image
 * question use these. Empty on a school whose registry declares no vision
 * model; callers fall back to a manual path rather than guessing.
 */
export function k9VisionModelNames(config: K9ModelsConfig): string[] {
  const runtime = config.runtime?.ollama;
  return ollamaNames(config, [runtime?.vision_model, ...(runtime?.vision_fallbacks ?? [])]);
}

export type K9ModelRegistry = {
  config: K9ModelsConfig;
  roots: Record<K9RootName, string>;
  configPath: string;
};

const COMPLETE_MARKER = ".k9_complete";

export function k9ModelsConfigPath(): string {
  const explicit = process.env["K9_MODELS_CONFIG"];
  if (explicit) return path.resolve(explicit);
  // `pnpm run start` runs from artifacts/api-server, two levels below the repo root.
  return path.resolve(process.cwd(), "..", "..", "config", "k9-models.json");
}

export function loadK9Models(configPath = k9ModelsConfigPath()): K9ModelRegistry {
  const config = JSON.parse(readFileSync(configPath, "utf8")) as K9ModelsConfig;
  const resolveRoot = (root: { path: string; env?: string }) =>
    path.resolve((root.env && process.env[root.env]) || root.path);
  return {
    config,
    roots: { base: resolveRoot(config.roots.base), k9: resolveRoot(config.roots.k9) },
    configPath,
  };
}

export function k9ModelPath(roots: Record<K9RootName, string>, entry: K9ModelEntry): string {
  return path.join(roots[entry.root ?? "k9"], ...entry.path.split("/"));
}

function matchesPattern(name: string, pattern: string): boolean {
  const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*");
  return new RegExp(`^${escaped}$`, "i").test(name);
}

export function k9ModelStatus(entry: K9ModelEntry, absPath: string): K9ModelStatus {
  if (!existsSync(absPath)) return "missing";
  switch (entry.kind) {
    case "file":
      return statSync(absPath).size > 0 ? "ready" : "partial";
    case "repo":
      return existsSync(path.join(absPath, ".git")) ? "ready" : "partial";
    case "files":
      return (entry.files ?? []).every((file) => existsSync(path.join(absPath, ...file.split("/"))))
        ? "ready"
        : "partial";
    default: {
      const names = readdirSync(absPath);
      const expected = (entry.expect ?? []).every((pattern) => names.some((name) => matchesPattern(name, pattern)));
      // Folders downloaded outside the K9 downloader have no completion marker:
      // they count once the expected files exist and nothing is still downloading.
      // A sharded safetensors model with every indexed shard on disk is complete
      // even if an earlier failed attempt left .incomplete pieces behind.
      if (entry.source.type === "external") {
        const settled = safetensorsIndexComplete(absPath) ?? !hasIncompleteDownloads(absPath);
        return names.length > 0 && expected && settled ? "ready" : "partial";
      }
      if (!names.includes(COMPLETE_MARKER)) return "partial";
      return expected ? "ready" : "partial";
    }
  }
}

/**
 * For a folder with model.safetensors.index.json: whether every shard it names
 * is present with at least the indexed number of bytes. null without an index.
 */
function safetensorsIndexComplete(dir: string): boolean | null {
  const indexPath = path.join(dir, "model.safetensors.index.json");
  if (!existsSync(indexPath)) return null;
  try {
    const index = JSON.parse(readFileSync(indexPath, "utf8")) as {
      metadata?: { total_size?: number };
      weight_map?: Record<string, string>;
    };
    const shards = [...new Set(Object.values(index.weight_map ?? {}))];
    if (shards.length === 0) return false;
    let bytes = 0;
    for (const shard of shards) {
      const file = path.join(dir, shard);
      if (!existsSync(file)) return false;
      bytes += statSync(file).size;
    }
    return bytes >= Number(index.metadata?.total_size ?? 0);
  } catch {
    return false;
  }
}

function hasIncompleteDownloads(absPath: string): boolean {
  const cache = path.join(absPath, ".cache", "huggingface", "download");
  if (!existsSync(cache)) return false;
  const pending = (dir: string, depth: number): boolean =>
    readdirSync(dir, { withFileTypes: true }).some((item) =>
      item.isDirectory()
        ? depth > 0 && pending(path.join(dir, item.name), depth - 1)
        : item.name.endsWith(".incomplete"),
    );
  return pending(cache, 3);
}
